conans/test/functional/subsystems_build_test.py

```diff
@@ -20,7 +25,129 @@ def test_cygwin_available(self):
         client.run_command('uname')
         assert "CYGWIN" in client.out
 
+    @pytest.mark.tool_msys2
+    @pytest.mark.tool_mingw32
+    def test_mingw32_available(self):
+        client = TestClient()
+        client.run_command('uname')
+        assert "MINGW32_NT" in client.out
+
+    @pytest.mark.tool_msys2
+    @pytest.mark.tool_mingw64
+    def test_mingw64_available(self):
+        client = TestClient()
+        client.run_command('uname')
+        assert "MINGW64_NT" in client.out
+
     def test_tool_not_available(self):
         client = TestClient()
         client.run_command('uname', assert_error=True)
         assert "'uname' is not recognized as an internal or external command" in client.out
+
+    makefile = textwrap.dedent("""
+        .PHONY: all
+        all: libhello.a app
+
+        app: main.o libhello.a
+        	$(CXX) $(CFLAGS) -o app main.o -lhello -L.
+
+        libhello.a: hello.o
+        	$(AR) rcs libhello.a hello.o
+
+        main.o: main.cpp
+        	$(CXX) $(CFLAGS) -c -o main.o main.cpp
+
+        hello.o: hello.cpp
+        	$(CXX) $(CFLAGS) -c -o hello.o hello.cpp
+        """)
+
+    def _build(self):
+        hello_h = gen_function_h(name="hello")
+        hello_cpp = gen_function_cpp(name="hello")
+        main_cpp = gen_function_cpp(name="main", includes=["hello"], calls=["hello"])
+
+        self.client.save({"Makefile": self.makefile,
+                          "hello.h": hello_h,
+                          "hello.cpp": hello_cpp,
+                          "main.cpp": main_cpp})
+
+        self.client.run_command("make")
+
+        self.client.run_command("app")
+
+    @pytest.mark.tool_msys2
+    def test_msys(self):
+        """
+        native MSYS environment, binaries depend on MSYS runtime (msys-2.0.dll)
+        posix-compatible, intended to be run only in MSYS environment (not in pure Windows)
+        """
+        # pacman -S gcc
+        self._build()
+
+        assert "__GNUC__" in self.client.out
```

Comment:
Check the ``check_exe_run`` helper instead

Reply:
used `check_exe_run` helper. I am not sure if I should add sub-system checks to the helper.